src/app/clusters/webrtc-transport-provider-server/webrtc-transport-provider-server.cpp

```diff
@@ -0,0 +1,501 @@
+
+        return UpsertResultEnum::kUpdated;
+    }
+
+    // If not found, insert a new entry
+    mCurrentSessions.push_back(session);
+
+    MatterReportingAttributeChangeCallback(AttributeAccessInterface::GetEndpointId().Value(), WebRTCTransportProvider::Id,
+                                           WebRTCTransportProvider::Attributes::CurrentSessions::Id);
+
+    return UpsertResultEnum::kInserted;
+}
+
+void WebRTCTransportProviderServer::RemoveSession(uint16_t sessionId)
+{
+    size_t originalSize = mCurrentSessions.size();
+
+    // Erase-Remove idiom
+    mCurrentSessions.erase(std::remove_if(mCurrentSessions.begin(), mCurrentSessions.end(),
+                                          [sessionId](const WebRTCSessionTypeStruct & s) { return s.id == sessionId; }),
+                           mCurrentSessions.end());
+
+    // If a session was removed, the size will be smaller.
+    if (mCurrentSessions.size() < originalSize)
+    {
+        // Notify the stack that the CurrentSessions attribute has changed.
+        MatterReportingAttributeChangeCallback(AttributeAccessInterface::GetEndpointId().Value(), WebRTCTransportProvider::Id,
+                                               WebRTCTransportProvider::Attributes::CurrentSessions::Id);
+    }
+}
+
+uint16_t WebRTCTransportProviderServer::GenerateSessionID()
+{
+    static uint16_t lastSessionId = 0;
+    uint16_t candidateId          = lastSessionId++;
+
+    // Handle wrap-around per spec
+    if (lastSessionId > kMaxSessionId)
+    {
+        lastSessionId = 0;
+    }
+
+    // Ensure uniqueness
+    while (FindSession(candidateId) != nullptr)
+    {
+        candidateId = lastSessionId++;
+        if (lastSessionId > kMaxSessionId)
+        {
+            lastSessionId = 0;
+        }
+    }
+
+    return candidateId;
+}
+
+// Command Handlers
+void WebRTCTransportProviderServer::HandleSolicitOffer(HandlerContext & ctx, const Commands::SolicitOffer::DecodableType & req)
+{
+    // According to the Matter WebRTC Transport Provider specification,
+    // at least one of VideoStreamID or AudioStreamID must be present.
+    if (!req.videoStreamID.HasValue() && !req.audioStreamID.HasValue())
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Protocols::InteractionModel::Status::ConstraintError);
+        return;
+    }
+
+    // Prepare arguments for Delegate::HandleSolicitOffer using OfferRequestArgs struct
+    Delegate::OfferRequestArgs args;
+    args.id                 = GenerateSessionID();
+    args.streamUsage        = req.streamUsage;
+    args.videoStreamID      = req.videoStreamID;
+    args.audioStreamID      = req.audioStreamID;
+    args.iceServers         = req.ICEServers;
+    args.iceTransportPolicy = req.ICETransportPolicy;
+    args.metadataOptions    = req.metadataOptions;
+    args.peerNodeID         = GetNodeIdFromCtx(ctx.mCommandHandler);
+    args.peerFabricIndex    = ctx.mCommandHandler.GetAccessingFabricIndex();
```

Comment:
So...  Does the spec intend for this command to only be used over CASE?  Because peerNodeID and peerFabricIndex might both be "not defined" here....

Followup to get that sorted out.

Reply:
Yes, [the spec say a Matter controller that has a CASE session with the camera is needed to coordinate the signaling phase between the camera and the viewer device in both signaling modes .](https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/0cfb12a2aa8fc4de7a66b8fc7ed675a364615b4c/src/app_clusters/webrtc.adoc#34-signaling-modes)

D

Reply:
Nothing in the command definitions says that the server is supposed to reject the command if not over CASE and, importantly, nothing says _how it should reject it_.

The spec needs to be fixed to define behavior here.

Reply:
https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/11320